Custom-Pylint-checks.md

```diff
@@ -0,0 +1,569 @@
+  * [Beyond hello world](#beyond-hello-world)
+* [Test custom Pylint checks](#test-custom-pylint-checks)
+  * [Design tests](#design-tests)
+  * [Pylint testing utilities](#pylint-testing-utilities)
+  * [Use testing utilities for hello world checkers](#use-testing-utilities-for-hello-world-checkers)
+    * [Tests for hello world AST checker](#tests-for-hello-world-ast-checker)
+    * [Tests for hello world token checker](#tests-for-hello-world-token-checker)
+    * [Tests for hello world raw checker](#tests-for-hello-world-raw-checker)
+  * [Run tests](#run-tests)
+
+## Introduction
+
+We use [Pylint](https://pylint.pycqa.org) to lint our Python code. Before writing a custom rule, you should check whether you can achieve your goals with [Pylint's built-in rules and configurations](https://pylint.pycqa.org/en/latest/technical_reference/features.html). If those are insufficient, you can follow this guide to write a custom Pylint checker.
+
+### Limits of linters
+
+Before we get too far into custom lint rules, let's make sure we understand what linters like ESLint are actually capable of doing.
+
+Consider two types things you might want to prohibit with a lint rule:
+
+* A property of the code's text. For example, you might want to ensure that all lines are at most 80 characters long.
+* A property of the code's behavior. For example, you might want to ensure that the code, when executed, does not print a particular string.
+
+Linters operate on the code as text, so you can write a lint rule that bans a textual property with perfect accuracy. However, any lint rule that tries to prohibit a behavior must be imperfect. To see this intuitively, consider the problem we'll approach below: writing a linter to prohibit programs that print `Hello, world!`. If our linter searched for the string `Hello, world!`, it would miss a program that printed `'Hello, ' + 'world!'`.
+
+If you are interested in computability theory, the technical description for this problem is that predicting a program's behavior is an undecidable problem. Look up the halting problem if you want to learn more.
+
+**If you want to prevent code that behaves a certain way, any linter you write must miss some code that performs the undesired behavior, raise an error on benign code, or both.**
+
+## Write checkers
+
+### Types of checkers
+
+Pylint supports three kinds of checkers:
+
+* **AST checkers** operate on an [abstract syntax tree (AST)](https://en.wikipedia.org/wiki/Abstract_syntax_tree) of the code that Pylint generates. The checker provides functions to be called as Pylint traverses the AST, and these functions can raise lint errors.
+
+* **Token checkers** operate on a list of the tokens that represents the code being linted. The tokens are of the same form as the tokens generated by [`tokenize.tokenize()`](https://docs.python.org/3/library/tokenize.html#tokenize.tokenize). To get familiar with how files get tokenized, put some simple Python code into `test.py` and then run:
```

Comment:
I think I'd suggest putting the stuff in "To get familiar..." onwards into the tokenize section below, rather than here (in the same way that you don't go into a lot of detail in this summary section about the AST representation of a program).

Reply:
done